File: src/main/kotlin/sunsetsatellite/vm/sunlite/SLTable.kt
package sunsetsatellite.vm.sunlite

import sunsetsatellite.lang.sunlite.Sunlite
import java.nio.file.Files.size

class SLTable(val sunlite: Sunlite) {
	private var map: MutableMap<AnySLValue, AnySLValue> = mutableMapOf()

	fun internal(): MutableMap<AnySLValue, AnySLValue> {
		return map
	}

	fun overwrite(m: MutableMap<AnySLValue, AnySLValue>): SLTable {
		this.map = m
		return this
	}

	fun set(index: AnySLValue, value: AnySLValue) {
		map[index] = value
	}

	fun get(index: AnySLValue): AnySLValue {
		return map[index] ?: SLNil
	}

	fun copy(): SLTable {
		return SLTable(sunlite).overwrite(map.mapKeys { it.key.copy() }.mapValues { it.value.copy() }.toMutableMap())
	}

	override fun toString(): String {
		return "<table of size ${map.size}>"
	}
}

File: src/main/csharp/SunliteSharp/Runtime/SLTable.cs
namespace SunliteSharp.Runtime;

public class SLTable(Sunlite sl)
{
    public AnySLValue this[AnySLValue key]
    {
        get => Internal[key];
        set => Internal[key] = value;
    }
    
    public Dictionary<AnySLValue, AnySLValue> Internal { get; private set; } = new();

    public SLTable Overwrite(Dictionary<AnySLValue, AnySLValue> m)
    {
        Internal = m;
        return this;
    }
    
    public SLTable Copy()
    {
        return new SLTable(sl).Overwrite(Internal.ToDictionary(kv => kv.Key.Copy(), kv => kv.Value.Copy()));
    }

    public override string ToString()
    {
        return $"<table of size {Internal.Count}>";
    }
}